File: src/main/scala/defo/ls/SRPath.scala
package defo.ls

import defo.core.Demand

class SRPath(val src: Int, val dest: Int, val bw: Double, maxMidpoints: Int) {
  private[this] val path = new Array[Int](maxMidpoints + 2)
  private[this] var size = 2
  path(0) = src
  path(1) = dest

  def length: Int = size
  
  def apply(i: Int): Int = path(i)
  
  def canInsert: Boolean = size < path.length

  def canRemove: Boolean = size > 2

  def insert(pos: Int, midpoint: Int): Unit = {
    var i = size
    while (i > pos) {
      path(i) = path(i - 1)
      i -= 1
    }
    path(i) = midpoint
    size += 1
  }

  // Compress if removing creates a double midpoint
  def remove(pos: Int): Unit = {
    var j = pos
    var i = pos + 1
    if (path(pos - 1) == path(pos + 1)) i += 1
    
    while (i < size) {
      path(j) = path(i)
      j += 1
      i += 1
    }
    size -= (i - j)
  }

  def substitute(pos: Int, midpoint: Int): Unit = {
    path(pos) = midpoint
  }

  def clear(): Unit = {
    path(1) = dest
    size = 2
  }
  
  override def toString: String = path.take(size).mkString("->")
}

object SRPath {
  def apply(d: Demand, maxMidpoints: Int): SRPath = {
    new SRPath(d.src, d.dest, d.bw, maxMidpoints)
  }
}

File: src/main/java/defo/ls/SRPath.java
package defo.ls;

public class SRPath {
  
  private int[] path;
  private int length;
  
  public SRPath(int src, int dest, int maxMidpoints) {
    this.path = new int[maxMidpoints + 2];
    this.length = 2;
    path[0] = src;
    path[1] = dest;
  }
  
  public int length() {
    return length;
  }
  
  public int src() {
    return path[0];
  }
  
  public int dest() {
    return path[length - 1];
  }
  
  public int get(int position) {
    return path[position];
  }
  
  public void insert(int position, int midpoint) {
    for (int i = length; i > position; i--) {
      path[i] = path[i - 1];
    }
    path[position] = midpoint;
    length++;
  }
  
  public void remove(int position) {
    int j = position;
    int i = j;
    if (path[position - 1] == path[position + 1]) {
      i++;
    }
    while (i < length) {
      path[j] = path[i];
      j++;
      i++;
    }
    length -= (i-j);
  }
  
  public void substitute(int position, int midpoint) {
    path[position] = midpoint;
  }
  
  public void clear() {
    path[1] = path[length - 1];
    length = 2;
  }
  
  @Override
  public String toString() {
    StringBuffer bf = new StringBuffer();
    for (int i = 0; i < length - 1; i++) {
      bf.append(path[i] + "->");
    }
    bf.append(path[length - 1]);
    return bf.toString();
  }
}
